src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -0,0 +1,313 @@
+    private readonly SequenceMessage _sequenceMessage = new(0);
+    private readonly Channel<int> _waitForAck = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = BoundedChannelFullMode.DropOldest });
+
+#if NET8_0_OR_GREATER
+    private readonly PeriodicTimer _timer = new(TimeSpan.FromSeconds(1));
+#else
+    private readonly TimerAwaitable _timer = new(TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+#endif
+    private readonly SemaphoreSlim _writeLock = new(1, 1);
+
+    private int _bufferIndex;
+    private long _totalMessageCount;
+    private bool _waitForSequenceMessage;
+
+    // Message IDs start at 1 and always increment by 1
+    private long _currentReceivingSequenceId = 1;
+    private long _latestReceivedSequenceId = long.MinValue;
+    private long _lastAckedId = long.MinValue;
+
+    private TaskCompletionSource<FlushResult> _resend = new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+    // TODO: pass in limits
+    public MessageBuffer(ConnectionContext connection, IHubProtocol protocol)
+    {
+        _buffer = new (SerializedHubMessage? Message, long SequenceId)[10];
+        for (var i = 0; i < _buffer.Length; i++)
+        {
+            _buffer[i].SequenceId = long.MinValue;
+        }
+        _connection = connection;
+        _protocol = protocol;
+
+        _resend.SetResult(new());
+
+#if !NET8_0_OR_GREATER
+        _timer.Start();
+#endif
+        _ = RunTimer();
+    }
+
+    private async Task RunTimer()
+    {
+        using (_timer)
+        {
+#if NET8_0_OR_GREATER
+            while (await _timer.WaitForNextTickAsync().ConfigureAwait(false))
+#else
+            while (await _timer)
+#endif
+            {
+                if (_lastAckedId < _latestReceivedSequenceId)
+                {
+                    // TODO: consider a minimum time between sending these?
+
+                    var sequenceId = _latestReceivedSequenceId;
+
+                    _ackMessage.SequenceId = sequenceId;
+
+                    await _writeLock.WaitAsync().ConfigureAwait(false);
+                    try
+                    {
+                        _protocol.WriteMessage(_ackMessage, _connection.Transport.Output);
+                        await _connection.Transport.Output.FlushAsync().ConfigureAwait(false);
+                        _lastAckedId = sequenceId;
+                    }
+                    finally
+                    {
+                        _writeLock.Release();
+                    }
+                }
+            }
+        }
+    }
+
+    // TODO: WriteAsync(HubMessage) overload, so we don't allocate SerializedHubMessage for messages that aren't going to be buffered
+    public async ValueTask<FlushResult> WriteAsync(SerializedHubMessage hubMessage, CancellationToken cancellationToken)
+    {
+        // TODO: Backpressure based on message count and total message size
+        if (_buffer[_bufferIndex].Message is not null)
```

Comment:
WriteAsync is always called in a lock by the calling code, so there will never be parallel WriteAsync calls, however we can't really express that here since this is a separate class.

We can add a comment to the method that it assumes the calling code is doing the needful.